src/dbnode/encoding/series_iterator_accumulator.go

```diff
@@ -0,0 +1,256 @@
+	err             error
+	firstNext       bool
+	closed          bool
+}
+
+// NewSeriesIteratorAccumulator creates a new series iterator.
+func NewSeriesIteratorAccumulator(
+	iter SeriesIterator,
+) (SeriesIteratorAccumulator, error) {
+	it := &seriesIteratorAccumulator{
+		id:              iter.ID(),
+		nsID:            iter.Namespace(),
+		seriesIterators: make([]SeriesIterator, 0, 2),
+	}
+
+	err := it.Add(iter)
+	if err != nil {
+		return nil, err
+	}
+
+	return it, nil
+}
+
+func (it *seriesIteratorAccumulator) Add(iter SeriesIterator) error {
+	if newNs := iter.Namespace(); !newNs.Equal(it.nsID) {
+		return fmt.Errorf("cannot add iterator with namespace %s to accumulator %s",
+			newNs.String(), it.nsID.String())
+	}
+
+	if !iter.Next() || !it.iters.push(iter) {
+		iter.Close()
+		return iter.Err()
+	}
+
+	iterStart := iter.Start()
+	if start := it.start; start.IsZero() || iterStart.Before(start) {
+		it.start = iterStart
+	}
+
+	iterEnd := iter.End()
+	if end := it.end; end.IsZero() || iterEnd.After(end) {
+		it.end = iterEnd
+	}
+
+	it.seriesIterators = append(it.seriesIterators, iter)
+	return nil
+
+}
+
+func (it *seriesIteratorAccumulator) ID() ident.ID {
+	return it.id
+}
+
+func (it *seriesIteratorAccumulator) Namespace() ident.ID {
+	return it.nsID
+}
+
+func (it *seriesIteratorAccumulator) Tags() ident.TagIterator {
+	it.err = errors.New("cannot get tags from series iterator accumulator")
+	return ident.EmptyTagIterator
+}
+
+func (it *seriesIteratorAccumulator) Start() time.Time {
+	return it.start
+}
+
+func (it *seriesIteratorAccumulator) End() time.Time {
+	return it.end
+}
+
+func (it *seriesIteratorAccumulator) Next() bool {
+	if !it.firstNext {
+		if !it.hasNext() {
+			return false
+		}
+
+		it.moveToNext()
+	}
+
+	it.firstNext = false
+	return it.hasNext()
+}
+
+func (it *seriesIteratorAccumulator) Current() (ts.Datapoint, xtime.Unit, ts.Annotation) {
+	return it.iters.current()
+}
+
+func (it *seriesIteratorAccumulator) Err() error {
+	if it.err != nil {
+		return it.err
+	}
+
+	for _, iter := range it.seriesIterators {
+		if err := iter.Err(); err != nil {
+			it.err = err
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (it *seriesIteratorAccumulator) Replicas() []MultiReaderIterator {
+	it.err = errors.New("cannot receive replicas on an iterator accumulator")
+	return []MultiReaderIterator{}
+}
+
+func (it *seriesIteratorAccumulator) Close() {
+	if it.isClosed() {
+		return
+	}
+	it.closed = true
+	if it.id != nil {
+		it.id.Finalize()
+		it.id = nil
+	}
+	if it.nsID != nil {
+		it.nsID.Finalize()
+		it.nsID = nil
+	}
+	for idx, iter := range it.seriesIterators {
+		iter.Close()
+		it.seriesIterators[idx] = nil
+	}
+	it.iters.reset()
+}
+
+func (it *seriesIteratorAccumulator) Reset(opts SeriesIteratorOptions) {
+	if len(it.seriesIterators) < 1 {
+		it.err = errors.New("cannot reset on empty series accumulator")
+		return
+	}
+
+	it.id = opts.ID
+	it.nsID = opts.Namespace
+	for _, it := range it.seriesIterators[1:] {
```

Comment:
Hm, why only close > 1?  Also should this check that each entry is not nil? (since in "Close()" they are nil'd out but slice is not shrunk `it.seriesIterators[idx] = nil`)

Reply:
So this would close every iterator except the first, which is instead reset, and set to be the `head` of the accumulator

Reply:
I guess I don't quite understand `Reset()`.

Is this meant to be called before or after "Add(....)"?

Usually "Reset(...)" means we pulled it from the pool and we're about to use it..